In [47]:
import pandas as pd
from nltk.tokenize import RegexpTokenizer, word_tokenize, TweetTokenizer
from nltk import FreqDist
import re
import string
from nltk.corpus import stopwords
import nltk
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\Gajavalli's\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


True

In [2]:
directory_path = "C:\\Users\\Gajavalli's\\Documents\\MS projects\\venky uk - sentiment analysis\\data\\"
cleaned_data_filename = 'news_dataset_cleaned.csv'

# Load Data

In [3]:
df = pd.read_csv(directory_path + cleaned_data_filename)

In [4]:
df.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject        date  \
0  Donald Trump just couldn t wish all Americans ...    News  2017-12-31   
1  House Intelligence Committee Chairman Devin Nu...    News  2017-12-31   
2  On Friday, it was revealed that former Milwauk...    News  2017-12-30   
3  On Christmas day, Donald Trump announced that ...    News  2017-12-29   
4  Pope Francis used his annual Christmas Day mes...    News  2017-12-25   

  label  title_len  text_len  caps_in_title  norm_caps_in_title  caps_in_text  \
0  fake         79      2893             11            0.139241           138   
1  fake         69      1898              8            0.115942            88   
2  fake         90      3597             15            0.166667           308   
3  fake         78      2774             19            0.243590           123   
4  fake         70      2346             11            0.157143            63   

   norm_caps_in_text  
0           0.047701  
1           0.046365  
2           0.085627  
3           0.044340  
4           0.026854

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38638 entries, 0 to 38637
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               38638 non-null  object 
 1   text                38638 non-null  object 
 2   subject             38638 non-null  object 
 3   date                38638 non-null  object 
 4   label               38638 non-null  object 
 5   title_len           38638 non-null  int64  
 6   text_len            38638 non-null  int64  
 7   caps_in_title       38638 non-null  int64  
 8   norm_caps_in_title  38638 non-null  float64
 9   caps_in_text        38638 non-null  int64  
 10  norm_caps_in_text   38638 non-null  float64
dtypes: float64(2), int64(4), object(5)
memory usage: 3.2+ MB


# Cleaning Title and Text Data

In [6]:
def concatination_of_strings(df, column):
    return [x for list_ in df[column].values for x in list_]

In [7]:
def find_strings(string_, regex):
    list_ = re.findall(regex, string_)
    return [s[0] for s in list_]

In [16]:
def frequencey_distribution_of_column(df, col):
    corpus_tokens = concatination_of_strings(df, col)
    corpus_freq_dist = FreqDist(corpus_tokens)
    print(f'Number of unique tokens in the corpus are {len(corpus_freq_dist)}')
    return corpus_freq_dist

In [9]:
def review_freq_dis(df, col, n):   
    corpus_freq_dist = frequencey_distribution_of_column(df, col)
    display(corpus_freq_dist.most_common(n))

In [10]:
def remove_punctuation(word_list, punctuation_list):
    return [w for w in word_list if w not in punctuation_list]

In [11]:
def remove_single_characters(word_list, exception_list):
    return [w for w in word_list if (len(w) > 1 or w in exception_list)]

In [12]:
def remove_words(word_list, words_to_remove):
    return [w for w in word_list if w not in words_to_remove]

## Token count distribution

In [13]:
tknzr = RegexpTokenizer(r'\w+|\$[\d\.]+|\([@\w\d]+\)')

In [14]:
df['text_tokens'] = df['text'].apply(tknzr.tokenize)

In [17]:
corpus_freq_dist = frequencey_distribution_of_column(df, 'text_tokens')

Number of unique tokens in the corpus are 152402


In [18]:
# lets look at the common, only once and less than 5 times of each token
corpus_freq_dist.most_common(150)

[('the', 771208),
 ('to', 457849),
 ('of', 372801),
 ('a', 339364),
 ('and', 336802),
 ('in', 282041),
 ('s', 198810),
 ('that', 195661),
 ('on', 162850),
 ('for', 144241),
 ('is', 138935),
 ('said', 122504),
 ('Trump', 121295),
 ('with', 98710),
 ('The', 98576),
 ('was', 96440),
 ('he', 95085),
 ('it', 88450),
 ('as', 82843),
 ('his', 80861),
 ('by', 79953),
 ('has', 78027),
 ('be', 71600),
 ('have', 70857),
 ('not', 68321),
 ('from', 67728),
 ('are', 60493),
 ('at', 59290),
 ('who', 58081),
 ('an', 56078),
 ('I', 51915),
 ('this', 50639),
 ('U', 50227),
 ('would', 49247),
 ('S', 47546),
 ('they', 47357),
 ('t', 44464),
 ('will', 42664),
 ('about', 41975),
 ('had', 40196),
 ('their', 38467),
 ('been', 36588),
 ('but', 36488),
 ('people', 34367),
 ('were', 34078),
 ('which', 33115),
 ('or', 32819),
 ('we', 32590),
 ('more', 31681),
 ('you', 31398),
 ('President', 31135),
 ('out', 29155),
 ('her', 28292),
 ('after', 28290),
 ('one', 27682),
 ('all', 26913),
 ('its', 26839),
 ('also', 26

In [23]:
print("Tokens used only once are:", len([w for w in corpus_freq_dist.most_common() if w[1] == 1]))

Tokens used only once are: 67886


In [24]:
print("Tokens used only once are:", len([w for w in corpus_freq_dist.most_common() if w[1] <= 5]))

Tokens used only once are: 107176


**Observations**:
1. The amount of words that are used only once or 5 or less times are small compared to the size of the corpus.

## Analysis of URLs

In [27]:
URL_REGEX = r"(?i)\b((?:https?://|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,4}/)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'\".,<>?«»“”‘’]))"

In [28]:
df['text_urls'] = df['text'].apply(lambda x: find_strings(x, URL_REGEX))

In [29]:
urls_in_text = concatination_of_strings(df, 'text_urls')

In [30]:
urls_in_text

['pic.twitter.com/4FPAe2KypA',
 'pic.twitter.com/XtZW5PdU2b',
 'pic.twitter.com/T2NY2psHCR',
 'https://t.co/zcbyc4Wp5b',
 'pic.twitter.com/fWfXsZupxy',
 '2017pic.twitter.com/ymsOBLjfxU',
 'pic.twitter.com/dWr5k8ZEZV',
 'pic.twitter.com/ulCFddhkdy',
 'https://t.co/Fg7VacxRtJ',
 'pic.twitter.com/5gEMcjQTbH',
 'https://t.co/zrWpyMXRcz',
 'pic.twitter.com/wiQSQNNzw0',
 'https://t.co/dkhw0AlHB4',
 'pic.twitter.com/oaZDT126B3',
 'https://t.co/ayBlGmk65Z',
 'pic.twitter.com/Z7dmyQ5smy',
 'pic.twitter.com/8TKtrMqRa1',
 'pic.twitter.com/hIxs3DciO8',
 'pic.twitter.com/E5bmcI83mU',
 'pic.twitter.com/a45En9Jwys',
 'pic.twitter.com/yLCBmhpNvG',
 'pic.twitter.com/3vMZUTEylx',
 'https://t.co/XrOvu32EV8',
 'pic.twitter.com/eMP9UX1bM8',
 'pic.twitter.com/XViyKFQCET',
 'https://t.co/HFYJRkefJ1',
 'https://t.co/65FhbQHuV4',
 'https://t.co/7lHYkIloyz',
 'https://t.co/g8SwgAKtfH',
 'https://t.co/9lCqpYujKN',
 'pic.twitter.com/NXEX9rGBgu',
 'pic.twitter.com/QePW9FtbSh',
 'pic.twitter.com/mUbKCIWGxB',
 'pic.

In [31]:
len(urls_in_text)

9648

**Observations**:
1. Number of URLs in the corpus are 9648
2. All the URLs looks like twitter links or images.

In [32]:
url_freq_dist = FreqDist(urls_in_text)

In [33]:
url_freq_dist.most_common(150)

[('bit.ly/2jBh4LU', 98),
 ('bit.ly/2jpEXYR', 98),
 ('connect.facebook.net/en_US/sdk.js#xfbml=1&#038;version=v2.3', 73),
 ('connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.3', 19),
 ('connect.facebook.net/en_GB/sdk.js#xfbml=1&#038;version=v2.3', 19),
 ('www.1100kfnx.com.LISTEN', 10),
 ('https://t.co/6OZtrfIwim', 9),
 ('pic.twitter.com/WYUnHjjUjg', 8),
 ('https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.11', 8),
 ('www.reuters.com/trump-effect', 8),
 ('https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.10', 7),
 ('https://t.co/4nxLipafWO', 7),
 ('connect.facebook.net/en_US/sdk.js#xfbml=1&amp;version=v2.3', 7),
 ('https://t.co/ifeSBlSZW7', 6),
 ('connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.10', 5),
 ('pic.twitter.com/QfefM8X2cW', 5),
 ('https://t.co/oaH1b92PFS', 5),
 ('pic.twitter.com/WIUTLOS4XD', 4),
 ('connect.facebook.net/en_GB/sdk.js#xfbml=1&version=v2.3', 4),
 ('https://t.co/hLWri23Lx7', 4),
 ('https://t.co/oLueAHCaUe', 4),
 ('pic.twitter.com/EHkQf

**Observations**:
1. The two URLs with bit.ly occured more times.
2. Passing more times of the URL will loss the importance of the model.
3. I will convert all the URLs to placeholder

In [34]:
df['clean_text'] = df['text'].apply(lambda x: re.sub(URL_REGEX, '{link}', x))

## Investigate URLs in title

In [36]:
# lets look at any URLs in the title.
df['title_urls'] = df['title'].apply(lambda x: find_strings(x, URL_REGEX))
urls_in_titles = concatination_of_strings(df, 'title_urls')
print(urls_in_titles)

[]


**Observations**:
1. There are no URLs in the title

## Analysis of twitter usernames in text

In [37]:
TWITTER_HANDLE_REGEX = r'(?<=^|(?<=[^\w]))(@\w{1,15})\b'

In [38]:
df['twitter_handles'] = df['clean_text'].apply(lambda x: re.findall(TWITTER_HANDLE_REGEX, x))
twitter_handles = concatination_of_strings(df, 'twitter_handles')
twitter_freq_dist = FreqDist(twitter_handles)

In [39]:
twitter_freq_dist.most_common(50)

[('@realDonaldTrump', 2997),
 ('@POTUS', 345),
 ('@21WIRE', 283),
 ('@FoxNews', 233),
 ('@HillaryClinton', 198),
 ('@seanhannity', 176),
 ('@CNN', 170),
 ('@nytimes', 126),
 ('@foxandfriends', 101),
 ('@elizabethforma', 98),
 ('@NBCNews', 93),
 ('@JordanUhl', 87),
 ('@tonyposnanski', 86),
 ('@ABC', 81),
 ('@AnnCoulter', 71),
 ('@bessbell', 69),
 ('@PressSec', 66),
 ('@FLOTUS', 65),
 ('@IvankaTrump', 59),
 ('@WalshFreedom', 53),
 ('@WhiteHouse', 52),
 ('@BernieSanders', 52),
 ('@KellyannePolls', 51),
 ('@PrisonPlanet', 49),
 ('@BraddJaffy', 49),
 ('@SpeakerRyan', 48),
 ('@realdonaldtrump', 48),
 ('@DonaldJTrumpJr', 47),
 ('@marcorubio', 46),
 ('@SarahPalinUSA', 45),
 ('@ABCPolitics', 44),
 ('@BarackObama', 43),
 ('@CNNPolitics', 39),
 ('@EricTrump', 39),
 ('@MSNBC', 39),
 ('@Morning_Joe', 39),
 ('@joshdcaplan', 39),
 ('@GOP', 38),
 ('@CBSNews', 38),
 ('@JackPosobiec', 38),
 ('@Cernovich', 37),
 ('@kylegriffin1', 36),
 ('@mike_pence', 36),
 ('@MatthewDicks', 36),
 ('@tedcruz', 36),
 ('@w

In [40]:
# total number of Twitter handles
len(twitter_handles)

22224

In [41]:
# number of unique Twitter handles (case sensitive)
len(twitter_freq_dist)

8162

**Observations**
1. realDonaldTrump is the most frequently occured twitter username in the dateset.
2. we can keep the actuals twitter username itself or we can replace all those with a generalzie twitter username.
   at this point of time will replace with general username.

In [42]:
df['clean_text'] = df['clean_text'].apply(lambda x: re.sub(TWITTER_HANDLE_REGEX, '@twitter-handle', x))

## Capitalization

In [43]:
# I want to keep the words that all characters in a word are capital as its and convert to lowercase of other words
# Words with length of one will be converted to lowercase without losing importance of them
def lower_unless_all_caps(string_):
    words = string_.split()
    processed_words = [w.lower() if not (w.isupper() and len(w) > 1) else w for w in words]
    return ' '.join(processed_words)

In [44]:
df['clean_text'] = df['clean_text'].apply(lower_unless_all_caps)
df['clean_title'] = df['title'].apply(lower_unless_all_caps)

## Numbers

In [45]:
# I would like to convert all the numbers with space because numbers mostly wont classify whether news is fake or true.
df['clean_text'] = df['clean_text'].apply(lambda x: re.sub(r'\d+', ' ', x))
df['clean_title'] = df['clean_title'].apply(lambda x: re.sub(r'\d+', ' ', x))

## Analysis of tokens from the cleaned_text

In [48]:
df['clean_text_tokens'] = df['clean_text'].apply(word_tokenize)

In [49]:
review_freq_dis(df, 'clean_text_tokens', 150)

Number of unique tokens in the corpus are 202031


[('the', 859816),
 (',', 779608),
 ('.', 545108),
 ('to', 458893),
 ('of', 373562),
 ('a', 352372),
 ('and', 345512),
 ('in', 297919),
 ('that', 203732),
 ('s', 199133),
 ('on', 166846),
 ('for', 147194),
 ('is', 139893),
 ('said', 120757),
 ('trump', 117656),
 ('he', 117174),
 ('it', 110523),
 ('with', 100695),
 ('was', 96714),
 ('as', 88045),
 ('his', 84136),
 ('by', 80892),
 ('has', 78136),
 ('not', 72446),
 ('be', 71204),
 ('have', 71038),
 ('’', 70279),
 ('from', 68567),
 (')', 68006),
 ('(', 67765),
 ('this', 62791),
 ('at', 62594),
 ('are', 61004),
 ('who', 58864),
 ('an', 57936),
 ('they', 57181),
 ('“', 53766),
 ('”', 53489),
 ('but', 50911),
 ('we', 50792),
 (':', 49644),
 ('would', 49324),
 ('i', 46700),
 ('president', 45310),
 ('U.S.', 43289),
 ('will', 43158),
 ('about', 42407),
 ('had', 40272),
 ('t', 40223),
 ('their', 39204),
 ('you', 37849),
 ('been', 36558),
 ('people', 35600),
 ('-', 34195),
 ('were', 34161),
 ('which', 33412),
 ('or', 33116),
 ('more', 32884),
 ('af

**Observations:**
1. I observe there are more stop words that occured more times in the text, those will be removed later stage.
2. Apart from the stop words observed that word 'reuters' occured at the starting of true news which will create a problem to ML model to overfit to this dataset.

## Removing (reuters) from news stories

In [50]:
df['clean_text'] = df['clean_text'].apply(lambda x: re.sub(r'\(reuters\)', ' ', x))

## Updating the  Tokens

In [51]:
df['clean_text_tokens'] = df['clean_text'].apply(word_tokenize)

In [52]:
review_freq_dis(df, 'clean_text_tokens', 150)

Number of unique tokens in the corpus are 202031


[('the', 859816),
 (',', 779608),
 ('.', 545108),
 ('to', 458893),
 ('of', 373562),
 ('a', 352372),
 ('and', 345512),
 ('in', 297919),
 ('that', 203732),
 ('s', 199133),
 ('on', 166846),
 ('for', 147194),
 ('is', 139893),
 ('said', 120757),
 ('trump', 117656),
 ('he', 117174),
 ('it', 110523),
 ('with', 100695),
 ('was', 96714),
 ('as', 88045),
 ('his', 84136),
 ('by', 80892),
 ('has', 78136),
 ('not', 72446),
 ('be', 71204),
 ('have', 71038),
 ('’', 70279),
 ('from', 68567),
 ('this', 62791),
 ('at', 62594),
 ('are', 61004),
 ('who', 58864),
 ('an', 57936),
 ('they', 57181),
 ('“', 53766),
 ('”', 53489),
 ('but', 50911),
 ('we', 50792),
 (':', 49644),
 ('would', 49324),
 (')', 46974),
 ('(', 46733),
 ('i', 46700),
 ('president', 45310),
 ('U.S.', 43289),
 ('will', 43158),
 ('about', 42407),
 ('had', 40272),
 ('t', 40223),
 ('their', 39204),
 ('you', 37849),
 ('been', 36558),
 ('people', 35600),
 ('-', 34195),
 ('were', 34161),
 ('which', 33412),
 ('or', 33116),
 ('more', 32884),
 ('af

**Observations**

Observed that reuters is no longer avaiable in the top 150 list.

## Removing puntuctions and single letter tokens from text

In [53]:
df['clean_text_tokens'] = df['clean_text_tokens'].apply(lambda x: remove_single_characters(x, ['i', '!']))

In [54]:
review_freq_dis(df, 'clean_text_tokens', 150)

Number of unique tokens in the corpus are 201940


[('the', 859816),
 ('to', 458893),
 ('of', 373562),
 ('and', 345512),
 ('in', 297919),
 ('that', 203732),
 ('on', 166846),
 ('for', 147194),
 ('is', 139893),
 ('said', 120757),
 ('trump', 117656),
 ('he', 117174),
 ('it', 110523),
 ('with', 100695),
 ('was', 96714),
 ('as', 88045),
 ('his', 84136),
 ('by', 80892),
 ('has', 78136),
 ('not', 72446),
 ('be', 71204),
 ('have', 71038),
 ('from', 68567),
 ('this', 62791),
 ('at', 62594),
 ('are', 61004),
 ('who', 58864),
 ('an', 57936),
 ('they', 57181),
 ('but', 50911),
 ('we', 50792),
 ('would', 49324),
 ('i', 46700),
 ('president', 45310),
 ('U.S.', 43289),
 ('will', 43158),
 ('about', 42407),
 ('had', 40272),
 ('their', 39204),
 ('you', 37849),
 ('been', 36558),
 ('people', 35600),
 ('were', 34161),
 ('which', 33412),
 ('or', 33116),
 ('more', 32884),
 ('after', 31621),
 ('she', 30880),
 ('one', 30119),
 ('if', 30084),
 ('her', 29018),
 ('state', 28335),
 ('out', 28326),
 ('all', 27997),
 ('what', 27418),
 ('its', 27295),
 ('also', 26967

## Tokenization of Titles

In [55]:
df['clean_title_tokens'] = df['clean_title'].apply(word_tokenize)

In [56]:
df.iloc[1000:1010]['clean_title_tokens']

1000    [stunning, new, poll, reveals, global, opinion...
1001    [former, GOP, rep, throws, support, behind, ob...
1002    [trump, moronically, claims, entire, russia, i...
1003    [WATCH, :, it, just, hit, a, trump, supporter,...
1004    [republicans, just, added, what, they, hated, ...
1005    [the, numbers, are, in, and, jon, ossoff, losi...
1006    [GOP, senator, lashes, into, kellyanne, conway...
1007    [cops, in, republican, senate, office, violent...
1008    [trump, is, giddy, about, his, upcoming, meeti...
1009    [catholic, sister, rips, ‘, pro-life, ’, repub...
Name: clean_title_tokens, dtype: object

In [57]:
review_freq_dis(df, 'clean_title_tokens', 150)

Number of unique tokens in the corpus are 31551


[('to', 14060),
 ('trump', 11939),
 (':', 10879),
 ('’', 10184),
 (',', 8742),
 ('in', 7536),
 ('of', 6083),
 ('for', 6043),
 ('on', 5630),
 ('the', 4638),
 ('s', 4564),
 ('U.S.', 4157),
 ("'s", 4120),
 ('(', 4101),
 (')', 4101),
 ('VIDEO', 3992),
 ('says', 3347),
 ('a', 3249),
 ('with', 3134),
 ('”', 3080),
 ('[', 3035),
 (']', 3034),
 ('and', 2942),
 ('“', 2847),
 ('‘', 2838),
 ("'", 2619),
 ('is', 2263),
 ('!', 2157),
 ('after', 2114),
 ('video', 1957),
 ('obama', 1807),
 ('house', 1794),
 ('at', 1774),
 ('as', 1763),
 ('over', 1694),
 ('from', 1551),
 ('he', 1533),
 ('his', 1528),
 ('by', 1484),
 ('about', 1455),
 ('white', 1377),
 ('new', 1367),
 ('it', 1354),
 ('will', 1312),
 ('not', 1309),
 ('WATCH', 1267),
 ('clinton', 1202),
 ('?', 1174),
 ('russia', 1142),
 ('just', 1133),
 ('be', 1122),
 ('president', 1117),
 ('hillary', 1058),
 ('bill', 1027),
 ('t', 1005),
 ('republican', 992),
 ('north', 986),
 ('korea', 912),
 ('that', 890),
 ('this', 888),
 ('senate', 878),
 ('out', 87

## Removing puntuctions and single letter tokens from title

In [58]:
df['clean_title_tokens'] = df['clean_title_tokens'].apply(lambda x: remove_single_characters(x, ['i', '!']))
review_freq_dis(df, 'clean_title_tokens', 150)

Number of unique tokens in the corpus are 31477


[('to', 14060),
 ('trump', 11939),
 ('in', 7536),
 ('of', 6083),
 ('for', 6043),
 ('on', 5630),
 ('the', 4638),
 ('U.S.', 4157),
 ("'s", 4120),
 ('VIDEO', 3992),
 ('says', 3347),
 ('with', 3134),
 ('and', 2942),
 ('is', 2263),
 ('!', 2157),
 ('after', 2114),
 ('video', 1957),
 ('obama', 1807),
 ('house', 1794),
 ('at', 1774),
 ('as', 1763),
 ('over', 1694),
 ('from', 1551),
 ('he', 1533),
 ('his', 1528),
 ('by', 1484),
 ('about', 1455),
 ('white', 1377),
 ('new', 1367),
 ('it', 1354),
 ('will', 1312),
 ('not', 1309),
 ('WATCH', 1267),
 ('clinton', 1202),
 ('russia', 1142),
 ('just', 1133),
 ('be', 1122),
 ('president', 1117),
 ('hillary', 1058),
 ('bill', 1027),
 ('republican', 992),
 ('north', 986),
 ('korea', 912),
 ('that', 890),
 ('this', 888),
 ('senate', 878),
 ('out', 876),
 ('state', 868),
 ('court', 863),
 ('china', 796),
 ('who', 788),
 ('him', 785),
 ('against', 782),
 ('up', 775),
 ('election', 763),
 ('you', 760),
 ('are', 758),
 ('has', 752),
 ('vote', 744),
 ('donald', 7

## Removing 's from the title and text

In [59]:
# Removing the 's from text and title because those 
# won't contribute much significance to classify fake or true news.
df['clean_title_tokens'] = df['clean_title_tokens'].apply(lambda x: remove_words(x, ["'s"]))
df['clean_text_tokens'] = df['clean_text_tokens'].apply(lambda x: remove_words(x, ["'s"]))

## Removing Day Words

In [60]:
# The day words and month names also does not contribute much classsifcaiton to the model \
# To get better generalzie model removing them from dataset
date_words = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 
              'saturday', 'sunday', 'january', 'february', 'march', 'april',
             'may', 'june', 'july', 'august', 'september', 'october',
             'november', 'december']
df['clean_title_tokens'] = df['clean_title_tokens'].apply(lambda x: remove_words(x, date_words))
df['clean_text_tokens'] = df['clean_text_tokens'].apply(lambda x: remove_words(x, date_words))

## Analyzing the Stop Words

In [ ]:
# Before removing stop words lets first compare most frequent words in the news titles and text with the NLTK English stopwords list

In [63]:
stop_words = stopwords.words('english')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Gajavalli's\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


In [64]:
display(stop_words)

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [65]:
most_freq_clean_text = [x[0] for x in list(frequencey_distribution_of_column(df, 'clean_text_tokens').most_common(150))]

Number of unique tokens in the corpus are 201920


In [66]:
def intersection(lst1, lst2):

    temp = set(lst2) 
    lst3 = [value for value in lst1 if value in temp] 
    return lst3

In [67]:
common_words = intersection(stop_words, most_freq_clean_text)

In [70]:
print("Common words:",len(common_words),",stop words:",len(stop_words))

Common words: 83 ,stop words: 179


**Observations**
1. There are 83 common stop words  and 97 stop words are not there in the NLTk stop word list.


In [72]:
def difference(lst1, lst2):

    temp = set(lst2) 
    lst3 = [value for value in lst1 if value not in temp] 
    return lst3

In [73]:
words_in_nltk_not_news = difference(stop_words, most_freq_clean_text)

In [74]:
words_in_news_not_nltk = difference(most_freq_clean_text, stop_words)

In [75]:
words_in_news_not_nltk

['said',
 'trump',
 'would',
 'president',
 'U.S.',
 'people',
 'one',
 'state',
 'also',
 'new',
 'donald',
 'house',
 'government',
 'states',
 'republican',
 'clinton',
 'twitter-handle',
 'obama',
 'could',
 'united',
 'told',
 'white',
 'campaign',
 'like',
 'last',
 'party',
 'two',
 'election',
 'time',
 'first',
 'former',
 'year',
 'country',
 'news',
 'even',
 '!',
 'many',
 'years',
 'security',
 'say',
 'percent',
 'republicans',
 'national',
 'court',
 'since',
 'political',
 'made',
 'law',
 'get',
 'make',
 'presidential',
 'hillary',
 'going',
 'administration',
 'senate',
 'media',
 'police',
 'russia',
 'democratic',
 'north',
 'american',
 'support',
 'week',
 'bill',
 'back',
 'including',
 'vote']

**Observaations**
1. Since i am doing sentiment analysis, removing the stop words may be indicative of fake or true news.
2. I am not removing at this point instead of that will compare the model performace with or without stopwords removed.

# Saving the Pre_processed_df

In [76]:
df.to_csv(directory_path + 'news_dataset_pre_processed.csv', index=False)